### ISSUE TITLE
compilation errors on macOS 10.15.4

### ISSUE THREAD

**Issue by mheese** [opendnssec/SoftHSMv2#543]
Hi there,

I just tried to update my SoftHSM to the 2.6.0 release, but I ran into compilation errors.

Here is how I ran `configure`.

```
./configure --enable-64bit --with-openssl=/usr/local/opt/openssl\@1.1 --with-sqlite3=/usr/local/opt/sqlite --with-migrate --with-botan=/usr/local/opt/botan --with-objectstore-backend-db --with-crypto-backend=openssl
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
configure: sysconfdir set to /etc
configure: localstate set to /var
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of gcc... gcc3
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking for pkg-config... /usr/local/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking if we can compile in 64-bit mode... yes
checking if we can run 64-bit programs... yes
checking whether g++ supports C++11 features by default... no
checking whether g++ supports C++11 features with -std=c++11... yes
checking for dlopen... yes
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking pthread.h usability... yes
checking pthread.h presence... yes
checking for pthread.h... yes
checking for FIPS 140-2 mode... no
checking for crypto backend... OpenSSL
checking what are the OpenSSL includes... -I/usr/local/opt/openssl@1.1/include
checking what are the OpenSSL libs... -L/usr/local/opt/openssl@1.1/lib -lcrypto
checking openssl/ssl.h usability... yes
checking openssl/ssl.h presence... yes
checking for openssl/ssl.h... yes
checking for BN_new in -lcrypto... yes
checking for OpenSSL version... >= 1.0.0
checking for OpenSSL ECC support... Found P256, P384, and P521
checking for OpenSSL EDDSA ED25519 support... Found ED25519
checking for OpenSSL EDDSA ED448 support... Found ED448
checking for OpenSSL GOST support... Cannot find GOST engine
checking OpenSSL EVP interface for AES key wrapping... RFC 3394 is supported
checking OpenSSL EVP interface for AES key wrapping with pad... RFC 5649 is supported
checking for ECC support... yes
checking for GOST support... no
checking for EDDSA support... yes
checking for non-paged memory for secure storage... enabled
checking sys/mman.h usability... yes
checking sys/mman.h presence... yes
checking for sys/mman.h... yes
checking the maximum size that may be locked into memory... unlimited
checking if building with softhsm2-migrate... yes
checking for sched_yield in -lrt... no
checking for sched_yield in -lposix4... no
checking sched.h usability... yes
checking sched.h presence... yes
checking for sched.h... yes
checking if building database object store backend... yes
checking if SQLite3 required... yes
checking for sqlite3... /usr/local/opt/sqlite/bin/sqlite3
checking what are the SQLite3 includes... -I/usr/local/opt/sqlite/include
checking what are the SQLite3 libs... -L/usr/local/opt/sqlite/lib -lsqlite3
checking sqlite3.h usability... yes
checking sqlite3.h presence... yes
checking for sqlite3.h... yes
checking for sqlite3_prepare_v2 in -lsqlite3... yes
checking for p11-kit integration... yes
checking where to install the p11-kit module... /usr/local/etc/pkcs11/modules
checking for cppunit-config... no
checking for cppunit... yes
checking for getpwuid_r... yes
checking build system type... x86_64-apple-darwin19.4.0
checking host system type... x86_64-apple-darwin19.4.0
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking how to convert x86_64-apple-darwin19.4.0 file names to x86_64-apple-darwin19.4.0 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin19.4.0 file names to toolchain format... func_convert_file_noop
checking for /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /bin/dd
checking how to truncate binary pipes... /bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... yes
checking for gcc option to produce PIC... -fno-common -DPIC
checking if gcc PIC flag -fno-common -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin19.4.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking whether the g++ linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking for g++ option to produce PIC... -fno-common -DPIC
checking if g++ PIC flag -fno-common -DPIC works... yes
checking if g++ static flag -static works... no
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin19.4.0 dyld
checking how to hardcode library paths into programs... immediate
checking that generated files are newer than configure... done
```

And then running `make` fails with a compilation error on `AESKey.cpp` in tests if I see that right:

```
make
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-recursive
Making all in src
Making all in lib
Making all in common
depbase=`echo Configuration.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT Configuration.lo -MD -MP -MF $depbase.Tpo -c -o Configuration.lo Configuration.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT Configuration.lo -MD -MP -MF .deps/Configuration.Tpo -c Configuration.cpp  -fno-common -DPIC -o .libs/Configuration.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT Configuration.lo -MD -MP -MF .deps/Configuration.Tpo -c Configuration.cpp -o Configuration.o >/dev/null 2>&1
depbase=`echo fatal.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT fatal.lo -MD -MP -MF $depbase.Tpo -c -o fatal.lo fatal.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT fatal.lo -MD -MP -MF .deps/fatal.Tpo -c fatal.cpp  -fno-common -DPIC -o .libs/fatal.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT fatal.lo -MD -MP -MF .deps/fatal.Tpo -c fatal.cpp -o fatal.o >/dev/null 2>&1
depbase=`echo log.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT log.lo -MD -MP -MF $depbase.Tpo -c -o log.lo log.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT log.lo -MD -MP -MF .deps/log.Tpo -c log.cpp  -fno-common -DPIC -o .libs/log.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT log.lo -MD -MP -MF .deps/log.Tpo -c log.cpp -o log.o >/dev/null 2>&1
depbase=`echo osmutex.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT osmutex.lo -MD -MP -MF $depbase.Tpo -c -o osmutex.lo osmutex.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT osmutex.lo -MD -MP -MF .deps/osmutex.Tpo -c osmutex.cpp  -fno-common -DPIC -o .libs/osmutex.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT osmutex.lo -MD -MP -MF .deps/osmutex.Tpo -c osmutex.cpp -o osmutex.o >/dev/null 2>&1
depbase=`echo SimpleConfigLoader.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT SimpleConfigLoader.lo -MD -MP -MF $depbase.Tpo -c -o SimpleConfigLoader.lo SimpleConfigLoader.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT SimpleConfigLoader.lo -MD -MP -MF .deps/SimpleConfigLoader.Tpo -c SimpleConfigLoader.cpp  -fno-common -DPIC -o .libs/SimpleConfigLoader.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT SimpleConfigLoader.lo -MD -MP -MF .deps/SimpleConfigLoader.Tpo -c SimpleConfigLoader.cpp -o SimpleConfigLoader.o >/dev/null 2>&1
depbase=`echo MutexFactory.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT MutexFactory.lo -MD -MP -MF $depbase.Tpo -c -o MutexFactory.lo MutexFactory.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT MutexFactory.lo -MD -MP -MF .deps/MutexFactory.Tpo -c MutexFactory.cpp  -fno-common -DPIC -o .libs/MutexFactory.o
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../crypto -I./../data_mgr -I./../pkcs11 -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT MutexFactory.lo -MD -MP -MF .deps/MutexFactory.Tpo -c MutexFactory.cpp -o MutexFactory.o >/dev/null 2>&1
/bin/sh ../../../libtool  --tag=CXX   --mode=link g++  -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden  -m64  -o libsofthsm_common.la  Configuration.lo fatal.lo log.lo osmutex.lo SimpleConfigLoader.lo MutexFactory.lo
libtool: link: ar cru .libs/libsofthsm_common.a .libs/Configuration.o .libs/fatal.o .libs/log.o .libs/osmutex.o .libs/SimpleConfigLoader.o .libs/MutexFactory.o
libtool: link: ranlib .libs/libsofthsm_common.a
libtool: link: ( cd ".libs" && rm -f "libsofthsm_common.la" && ln -s "../libsofthsm_common.la" "libsofthsm_common.la" )
Making all in crypto
Making all in test
make[5]: Nothing to be done for `all'.
depbase=`echo AESKey.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../..  -I./.. -I./../common -I./../data_mgr -I./../pkcs11 -I/usr/local/opt/openssl@1.1/include   -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT AESKey.lo -MD -MP -MF $depbase.Tpo -c -o AESKey.lo AESKey.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../.. -I./.. -I./../common -I./../data_mgr -I./../pkcs11 -I/usr/local/opt/openssl@1.1/include -m64 -g -O2 -Wall -Wextra -std=c++11 -fvisibility=hidden -MT AESKey.lo -MD -MP -MF .deps/AESKey.Tpo -c AESKey.cpp  -fno-common -DPIC -o .libs/AESKey.o
In file included from AESKey.cpp:37:
In file included from ./CryptoFactory.h:40:
In file included from ./AsymmetricAlgorithm.h:37:
In file included from ./AsymmetricKeyPair.h:38:
./PublicKey.h:64:23: warning: 'constexpr' non-static member function will not be
      implicitly 'const' in C++14; add 'const' to avoid a change in behavior
      [-Wconstexpr-not-const]
        constexpr PublicKey& operator=(const PublicKey&) { return *this; }
                             ^
                                                         const
./PublicKey.h:64:60: error: binding value of type 'const PublicKey' to reference
      to type 'PublicKey' drops 'const' qualifier
        constexpr PublicKey& operator=(const PublicKey&) { return *this; }
                                                                  ^~~~~
./PublicKey.h:64:23: error: no return statement in constexpr function
        constexpr PublicKey& operator=(const PublicKey&) { return *this; }
                             ^
In file included from AESKey.cpp:37:
In file included from ./CryptoFactory.h:40:
In file included from ./AsymmetricAlgorithm.h:37:
In file included from ./AsymmetricKeyPair.h:39:
./PrivateKey.h:71:24: warning: 'constexpr' non-static member function will not
      be implicitly 'const' in C++14; add 'const' to avoid a change in behavior
      [-Wconstexpr-not-const]
        constexpr PrivateKey& operator=(const PrivateKey&) { return *this; }
                              ^
                                                           const
./PrivateKey.h:71:62: error: binding value of type 'const PrivateKey' to
      reference to type 'PrivateKey' drops 'const' qualifier
        constexpr PrivateKey& operator=(const PrivateKey&) { return *this; }
                                                                    ^~~~~
./PrivateKey.h:71:24: error: no return statement in constexpr function
        constexpr PrivateKey& operator=(const PrivateKey&) { return *this; }
                              ^
2 warnings and 4 errors generated.
make[5]: *** [AESKey.lo] Error 1
make[4]: *** [all-recursive] Error 1
make[3]: *** [all-recursive] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

**Comment by mheese** [comment]
Looks related to https://github.com/opendnssec/SoftHSMv2/issues/538

**Comment by rijswijk** [comment]
Will be fixed with the merge of #539 (which will be included in 2.6.1)
